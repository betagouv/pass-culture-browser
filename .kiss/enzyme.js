import React from 'react'
import { shallow } from 'enzyme'

import MyComponent from '../MyComponent'

describe('src | components | MyComponent', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      const props = { errors: {} }
      const wrapper = shallow(<MyComponent {...props} />)
      expect(wrapper).toBeDefined()
      expect(wrapper).toMatchSnapshot()
    })
  })
})
