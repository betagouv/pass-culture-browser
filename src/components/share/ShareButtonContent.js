import PropTypes from 'prop-types'
import React from 'react'

import CopyToClipboardButton from './CopyToClipboardButton'
import MailToLink from '../layout/MailToLink'
import { openSharePopin, closeSharePopin } from '../../reducers/share'

export const getCopyToClipboardButton = (url, onClick) => (
  <CopyToClipboardButton
    className="py12 is-bold fs14"
    key="CopyToClipboard"
    onClick={onClick}
    value={url}
  />
)
export const getMailToLinkButton = (email, headers) => (
  <MailToLink
    className="no-underline is-block is-white-text py12 is-bold fs14"
    email={email}
    headers={headers}
    key="MailToLink"
  >
    <span>{'Envoyer par e-mail'}</span>
  </MailToLink>
)

export const getCloseButton = onClose => (
  <button
    className="no-border no-background no-outline is-block py12 is-bold fs14"
    key="closeButton"
    onClick={onClose}
    type="button"
  >
    <span>{'Fermer'}</span>
  </button>
)

class ShareButtonContent extends React.PureComponent {
  constructor(props) {
    super(props)
    this.state = { iscopied: false }
  }

  onCopyHandler = status => {
    this.setState({ iscopied: status }, () => {
      this.handleOnClickShare()
    })
  }

  onCloseHandler = () => {
    const { dispatch } = this.props
    dispatch(closeSharePopin())
    this.setState({ iscopied: false })
  }

  handleOnClickShare = () => {
    const { dispatch, email, offerName, text, url } = this.props

    try {
      const nativeOptions = { text: `${text}\n`, title: text, url }
      return navigator
        .share(nativeOptions)
        .then(() => {})
        .catch(() => {})
    } catch (err) {
      const headers = {
        body: `${text}\n\n${url}`,
        subject: text,
      }
      const popinOptions = {
        buttons: [
          getCopyToClipboardButton(url, this.onCopyHandler),
          getMailToLinkButton(email, headers),
        ],
        text: 'Comment souhaitez-vous partager cette offre ?',
        title: offerName,
      }

      const { iscopied } = this.state

      if (iscopied) {
        popinOptions.text = 'Le lien a bien été copié'
        popinOptions.buttons = [getCloseButton(this.onCloseHandler)]
      }

      return dispatch(openSharePopin(popinOptions))
    }
  }

  render() {
    const { id, offerName, url } = this.props
    const isDisabled = !offerName || !url
    return (
      <button
        className="no-border no-background"
        disabled={isDisabled}
        id={id}
        onClick={this.handleOnClickShare}
        type="button"
      >
        <span
          aria-hidden
          className="icon-ico-share"
          title="Partager cette offre"
        />
      </button>
    )
  }
}

ShareButtonContent.defaultProps = {
  email: null,
  id: 'verso-share-button',
  offerName: null,
  text: null,
  url: null,
}

ShareButtonContent.propTypes = {
  dispatch: PropTypes.func.isRequired,
  email: PropTypes.string,
  id: PropTypes.string,
  offerName: PropTypes.string,
  text: PropTypes.string,
  url: PropTypes.string,
}

export default ShareButtonContent
