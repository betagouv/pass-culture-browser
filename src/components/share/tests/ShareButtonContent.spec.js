import React from 'react'
import { shallow } from 'enzyme'

import ShareButtonContent, {
  getCopyToClipboardButton,
  getMailToLinkButton,
  getCloseButton,
} from '../ShareButtonContent'

const dispatchMock = jest.fn()

describe('src | components | share | getCopyToClipboardButton', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      // given
      const props = {
        onClick: () => {},
        url: 'http://foo.com',
      }
      // when
      const wrapper = shallow(getCopyToClipboardButton(...props))
      // then
      expect(wrapper).toBeDefined()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('On click copy', () => {
    it('should call function on click', () => {
      // given
      const onClick = jest.fn()
      const props = {
        onClick,
        url: 'http://foo.com',
      }
      // when
      const wrapper = shallow(getCopyToClipboardButton(...props))
      const button = wrapper.find('button').simulate('click')

      // then
      expect(wrapper).toBeDefined()
      expect(button.length).toBe(1)
      // FIXME : can't understand why test does not pass :'(
      //    expect(onClick).toHaveBeenCalled()
    })
  })
})

describe('src | components | share | getMailToLinkButton', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      // given
      const props = {
        email: 'foo@bar.com',
        headers: {
          subject: 'Fake title',
          url: 'http://www.fake-url.com',
        },
      }
      // when
      const wrapper = shallow(getMailToLinkButton(...props))
      // then
      expect(wrapper).toBeDefined()
      expect(wrapper).toMatchSnapshot()
    })
  })
})

describe('src | components | share | getCloseButton', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      // given
      const props = {
        onClose: () => {},
      }
      // when
      const wrapper = shallow(getCloseButton(...props))
      // then
      expect(wrapper).toBeDefined()
      expect(wrapper).toMatchSnapshot()
    })
  })
})

describe('src | components | share | ShareButtonContent', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      // given
      const props = {
        dispatch: dispatchMock,
        title: 'Fake title',
        url: 'http://www.fake-url.com',
      }

      // when
      const wrapper = shallow(<ShareButtonContent {...props} />)

      // then
      expect(wrapper).toBeDefined()
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('functions', () => {
    describe('onClickShare', () => {
      describe('when there is not native sharing', () => {
        it('should call dispatch with good action parameters', () => {
          // given
          const props = {
            dispatch: dispatchMock,
            email: 'foo@bar.com',
            title: 'Fake title',
            url: 'http://www.fake-url.com',
          }

          // when
          const wrapper = shallow(<ShareButtonContent {...props} />)
          wrapper.find('button').simulate('click')
          const expected = {
            options: {
              buttons: expect.any(Array),
              text: 'Comment souhaitez-vous partager cette offre ?',
              title: 'Fake title',
              url: 'http://www.fake-url.com',
            },
            type: 'TOGGLE_SHARE_POPIN',
          }

          // then
          expect(dispatchMock).toHaveBeenCalledWith(expected)
        })
      })
    })
  })
})
