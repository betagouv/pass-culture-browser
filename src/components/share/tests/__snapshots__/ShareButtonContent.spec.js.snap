// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | share | ShareButtonContent snapshot should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ShareButtonContent
    dispatch={[MockFunction]}
    email={null}
    id="verso-share-button"
    offerName="Fake offer name"
    text="Fake text"
    url="http://www.fake-url.com"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <span
        aria-hidden={true}
        className="icon-ico-share"
        title="Partager cette offre"
      />,
      "className": "no-border no-background",
      "disabled": false,
      "id": "verso-share-button",
      "onClick": [Function],
      "type": "button",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-hidden": true,
        "className": "icon-ico-share",
        "title": "Partager cette offre",
      },
      "ref": null,
      "rendered": null,
      "type": "span",
    },
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <span
          aria-hidden={true}
          className="icon-ico-share"
          title="Partager cette offre"
        />,
        "className": "no-border no-background",
        "disabled": false,
        "id": "verso-share-button",
        "onClick": [Function],
        "type": "button",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-hidden": true,
          "className": "icon-ico-share",
          "title": "Partager cette offre",
        },
        "ref": null,
        "rendered": null,
        "type": "span",
      },
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`src | components | share | getCloseButton snapshot should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <button
    className="no-border no-background no-outline is-block py12 is-bold fs14"
    onClick={[Function]}
    type="button"
  >
    <span>
      Fermer
    </span>
  </button>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "closeButton",
    "nodeType": "host",
    "props": Object {
      "children": <span>
        Fermer
      </span>,
      "className": "no-border no-background no-outline is-block py12 is-bold fs14",
      "onClick": [Function],
      "type": "button",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Fermer",
      },
      "ref": null,
      "rendered": "Fermer",
      "type": "span",
    },
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "closeButton",
      "nodeType": "host",
      "props": Object {
        "children": <span>
          Fermer
        </span>,
        "className": "no-border no-background no-outline is-block py12 is-bold fs14",
        "onClick": [Function],
        "type": "button",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Fermer",
        },
        "ref": null,
        "rendered": "Fermer",
        "type": "span",
      },
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;

exports[`src | components | share | getCopyToClipboardButton snapshot should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CopyToClipboardButton
    className="py12 is-bold fs14"
    onClick={[Function]}
    value="http://foo.com"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <button
        className="no-border no-background no-outline is-block py12 is-bold fs14"
        type="button"
      >
        <span>
          Copier le lien
        </span>
      </button>,
      "onCopy": [Function],
      "options": undefined,
      "text": "http://foo.com",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <span>
          Copier le lien
        </span>,
        "className": "no-border no-background no-outline is-block py12 is-bold fs14",
        "type": "button",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Copier le lien",
        },
        "ref": null,
        "rendered": "Copier le lien",
        "type": "span",
      },
      "type": "button",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <button
          className="no-border no-background no-outline is-block py12 is-bold fs14"
          type="button"
        >
          <span>
            Copier le lien
          </span>
        </button>,
        "onCopy": [Function],
        "options": undefined,
        "text": "http://foo.com",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <span>
            Copier le lien
          </span>,
          "className": "no-border no-background no-outline is-block py12 is-bold fs14",
          "type": "button",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Copier le lien",
          },
          "ref": null,
          "rendered": "Copier le lien",
          "type": "span",
        },
        "type": "button",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`src | components | share | getMailToLinkButton snapshot should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MailToLink
    className="no-underline is-block is-white-text py12 is-bold fs14"
    email="foo@bar.com"
    headers={
      Object {
        "subject": "Fake title",
        "url": "http://www.fake-url.com",
      }
    }
    obfuscate={false}
  >
    <span>
      Envoyer par e-mail
    </span>
  </MailToLink>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <span>
        Envoyer par e-mail
      </span>,
      "className": "no-underline is-block is-white-text py12 is-bold fs14",
      "href": "mailto:foo@bar.com?subject=Fake%20title&url=http%3A%2F%2Fwww.fake-url.com",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Envoyer par e-mail",
      },
      "ref": null,
      "rendered": "Envoyer par e-mail",
      "type": "span",
    },
    "type": "a",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <span>
          Envoyer par e-mail
        </span>,
        "className": "no-underline is-block is-white-text py12 is-bold fs14",
        "href": "mailto:foo@bar.com?subject=Fake%20title&url=http%3A%2F%2Fwww.fake-url.com",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Envoyer par e-mail",
        },
        "ref": null,
        "rendered": "Envoyer par e-mail",
        "type": "span",
      },
      "type": "a",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
